src/validators/datetime-or-date.test.ts

```diff
@@ -0,0 +1,107 @@
+import { NotDatetimeOrDateFail, RequiredFail } from '../errors'
+import {
+  DateTimeOrDate,
+  DateTimeOrDateValidator,
+  OptionalDateTimeOrDate,
+  RequiredDateTimeOrDate,
+  validateDateTimeOrDate
+} from './datetime-or-date'
+
+describe.each([false, true])('DateTime (optimize: %s)', optimize => {
+  describe('validateDateTimeOrDate', () => {
+    it('requires value to be an RFC 3339 timestamp', () => {
+      expect(validateDateTimeOrDate('2018-08-06T13:37:00Z')).toStrictEqual([])
+      expect(validateDateTimeOrDate(new Date('2018-08-06T13:37:00Z'))).toStrictEqual([])
+    })
+  })
+
+  describe('DateTimeOrDateValidator', () => {
+    it.skip('should generate validation code and give same result', () => {
```

Comment:
Why is this skipped?

Reply:
> What is supposed to happen with the optimize TODOs?

You don't have to implement an optimised version as it will fallback to naiv implementation that just calls the validate function, and it might not make sense here when we get unions.